woocommerce/woocommerce-ios
# Order Fulfillment: Add custom providers at the beginning of the list of shipment providers

Implements the first part of #847 : add a "Custom" section at the beginning of the list of shipment providers.

I split #847 in two because it seems like adding the store's country providers to the beginning of the list right require (as far as I can tell) touching the Storage and Yosemite layers. Learning from past mistakes, it seems like it might be better to go merging improvements incrementally than attacking everything in a huge PR.

## Changes
All changes are isolated in `ShipmentProvidersViewController` and `ShippingProvidersViewModel`. 

As mentioned [in slack](p1556518345079200-slack-woo-mobile), I didn't manage to find a way to sort the list of providers the way it is required for this ticket if I add the "Custom" group to the CoreData store manually. 

The only way I found to be able to sort the list as required is encapsulating the data model in the view model and creating a "Custom" section on the fly.

## Testing
- Checkout the branch, run the tests
- Navigate to an order > Fulfill order > Add Tracking
- Tap the first row (Shipping Provider)

It should look like this:
![Simulator Screen Shot - iPhone 8 - 2019-04-29 at 16 25 17](https://user-images.githubusercontent.com/2722505/56884127-65afb580-6a9b-11e9-8c79-c37eb5a11688.png)

Notice the first section (Custom). Tapping on it should navigate to the screen to introduce a custom shipment provider. Tapping any other provider should mark it as selected.


Update release notes:
